### PR TITLE
Enhance configuration system

### PR DESCRIPTION
**Pull request by Lyrkan**
Not sure about that one but the idea would be to use [node-convict](https://github.com/mozilla/node-convict) to handle settings.

It'd allow us to:
* validate settings (see schema in `configuration.js`) ;
* use environment variables (e.g. `PORT=8001 node app.js `) ;
* use command line arguments (e.g. `mockiji --port 8001`) ;
* specify if a config file should be loaded to override the default configuration (e.g. `mockiji --config-file dev.json`) and thus remove the need of having a `config/env.json`file.
* call mockiji from any JS file (e.g. `(new require('mockiji')({configuration: {...})).start();`)

Some breaking changes though:
* No more `config/default.json` and `config/env.json` files
* Renamed settings:
  * `listen_to_these_api_base_urls.api` to `api_base_path`
  * `mock_file_not_found_http_code` to `http_codes.mock_file_not_found`
  * `mock_file_invalid_http_code` to `http_codes.mock_file_invalid`
* Removed setting:
  * `path` from `env.json` (not needed anymore thanks to the `--config-file` argument)
* Moved settings:
  * `env` is now directly set in the default settings

TODO :
- [x] Use node-convict
- [x] Move content of the 'api/' folder to the root of the repository
- [x] Edit README.md
- [x] Add a proper documentation and type check for every field of the schema
- [x] Create a `.npmignore` file to exclude examples and files not used in a npm context
- [ ] Check if everything works properly with Maestro

### CODE REVIEW

File: api/processes.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "apps" : [{
     "name"               : "mockiji-api",
-    "script"             : "app.js",
+    "script"             : "app.js --config-file ../../mocks/data/dev.json",
```
Comment (ldiringer) [inline api/processes.json]: This path does not exists in this standalone context. 
I would like people to be able to clone this repository and launch it without editing anything and they would be able to use the "/mocks/api-simple-library" as described in the "Get started with default configuration" part from the README.

Also, I think it's important to be able to override this default configuration from an untracked file.
Before you're merge-request, I was thinking about an idea of creating a env.json.dist file and use it to create a env.json file if this last file didn't exist.

Comment (Lyrkan) [reply api/processes.json]: Yeah, I kinda hesitated before using that path, but that was the one referenced in `automation/bin/install` before.

I'm gonna move it back to the `automation/bin/install` and modify the README to include the `--api-base-path` argument so it uses the `api-simple-library` example.

Now concerning why I dropped the `env.json` file.
I may be wrong but I think the most logical way of using Mockiji would be to add it as a dev-dependency in a project and start it from a npm script (or gulp, webpack, ...):

```
{
  "name": "my-app",
  "scripts": {
    (...),
    "mocks": "mockiji --config-file ./mockiji.conf"
  },
  (...)
  "devDependencies": {
    "mockiji": "latest"
  }
}
```

 If that's that case, people are not going to edit an `env.json` file but rather reference their own config file.

**Comment by Lyrkan** [comment]
Updated with what we talked about earlier + merged `master`.

Maybe some new breaking changes though:
* paths from the configuration are resolved from the current directory: it makes a bit more sense since if someone types `mockiji --api-base-path ./mocks`, he'll probably expect `mocks/` to be the one of the folder he's currently in.
* default values for `--api-base-path` and `--logger-path` are now `./mocks` and `./logs/api-mockiji.log`: it was a bit weird for the `mockiji` command to look for mocks and put logs in the parent directory

We'll have to test it using Maestro to see if it works properly.

Also, @ldiringer: do you have a clear explanation about what the `--authorization_token` parameter does? I'm not sure what to put there.

**Comment by ldiringer** [comment]
The `authorization_token` may have to be renamed. It is to simulate an "anonymous" request (containing no user id), identified by the `Authorization` request header.

Example for the configuration:
```json
"authorization_token": [
    {
      "regexp": "/api/v1/users/(-)",
      "token_replacement_key": "user_id"
    }
]
```

For this to work, the second word from the `Authorization` request header value has to be a base64 encoded JSON object containing a "user_id" key which correspond to a string concatenable value (string, int, etc.).
```json
{
  "user_id": 567,
  "session_creation_date": "2017-04-03T20:12:45.000Z"
}
```

Now, for any request containing `/api/v1/users/-`, the dash `-` will be replaced by the token `user_id` value `567`.  

Example with the above token and configuration:  `GET /api/v1/users/-/books`
Mockiji will search for file matching the request `GET /api/v1/users/567/books`

**Pull request merged**